### ISSUE TITLE
InputfieldFile does not abide by noAjax setting when used in Page Edit

### ISSUE THREAD

**Issue by Toutouwai** [processwire/processwire-issues#1836]
### Short description of the issue

InputfieldFile has a `noAjax` setting, described as "Set to true or 1 to disable ajax uploading".

But when InputfieldFile appears in Page Edit, ajax uploading is always used even if `noAjax` is set to true or 1. This seems to be related to the presence of the `#PageIDIndicator` element in Page Edit.

In my case I want to add an InputfieldFile to the Page Edit form using a hook and have it work in the form the same way it would in any other PW form besides Page Edit. There will be no corresponding Field object (FieldtypeFile) for this inputfield and any file selected in this inputfield does not need to be saved anywhere. But the forced ajax uploading prevents a file from being selected in the inputfield.

### Expected behavior

`noAjax` disables ajax uploading.

### Setup/Environment

* ProcessWire version: 3.0.230

**Comment by ryancramerdesign** [comment]
@Toutouwai That noAjax mode was built for situations outside the page editor, like if another Process module needs to provide file uploads, but doesn't have the ajax handler that's built into ProcessPageEdit. I'm not sure that this ajax-disabled mode would work in the page editor, but it could be worth a try. Edit that InputfieldFile.js and InputfieldImage.js files (whichever you  need) to force them to use the old/non-ajax method. I think this would be around line 565 of InputfieldFile.js and line 41 of InputfieldImage.js (just add a `return false`). If you find it still works, let me know, and we can always add a check for the module's noAjax property on the JS side.

**Comment by Toutouwai** [comment]
@ryancramerdesign, I tried both InputfieldFile and InputfieldImage in Page Edit with the ajax option manually disabled and although there are UI details that could be improved they do both work in that the added files/images get saved to the page after submit.

**Comment by matjazpotocnik** [comment]
@ryancramerdesign 

Confirmed on the current `dev` branch tip (3.0.272 dev). Reproduced — `InputfieldFile::$noAjax` is documented as *"Set to true or 1 to disable ajax uploading"* and is honored everywhere in the PHP side (`InputfieldFile.module` lines 757, 908, 1285, 1314 all check `!$this->noAjax` before doing ajax work), but the client-side initializer in `InputfieldFile.js` ignores it entirely when the inputfield is rendered in Page Edit.

## Root cause

The decision to enable HTML5 ajax upload is **global, not per-inputfield**. In `wire/modules/Inputfield/InputfieldFile/InputfieldFile.js` (lines 606–613 on the dev tip):

```js
var allowAjax = false;
if (window.File && window.FileList && window.FileReader 
        && ($("#PageIDIndicator").length > 0 || $('.InputfieldAllowAjaxUpload').length > 0)) {  
    InitHTML5('');  
    allowAjax = true;
} else {
    InitOldSchool();
}
```

The comment even says it out loud: *"#PageIDIndicator.length indicates PageEdit, which we're limiting AjaxUpload to since only ProcessPageEdit has the ajax handler"*.

The check is: **if we're in Page Edit (`#PageIDIndicator` exists) OR any field on the page has the `InputfieldAllowAjaxUpload` class, then ALL `InputfieldFile` instances get HTML5 ajax init**. There's no per-field escape hatch — `noAjax=true` on a single field is silently ignored.

`InitHTML5('')` then iterates every `.InputfieldFileUpload` on the page and binds a `change` listener directly on each file input (line 461):

```js
filesUpload.addEventListener("change", function(evt) {
    traverseFiles(this.files);
    evt.preventDefault();
    evt.stopPropagation();
    this.value = '';     // ← this is what "prevents a file from being selected"
}, false);
```

That `this.value = ''` line clears the file input immediately after the user picks a file, so when the form is submitted the file isn't there. `evt.stopPropagation()` also prevents `InitOldSchool`'s delegated change handler (which would normally update the file label) from firing — but `InitOldSchool` is only called in the `else` branch, so on a Page Edit form it isn't even bound.

This is why the reporter can't even pick a file in their hook-added `InputfieldFile`: the HTML5 change handler grabs the file, attempts the ajax upload (which fails because there's no FieldtypeFile backend or no matching ajax endpoint for the hook-added field), and then clears the input.

## Proposed fix
```diff
--- a/wire/modules/Inputfield/InputfieldFile/InputfieldFile.module
+++ b/wire/modules/Inputfield/InputfieldFile/InputfieldFile.module
@@ -783,6 +783,13 @@ class InputfieldFile extends Inputfield implements InputfieldItemList, Inputfiel
 		$config = $this->wire()->config;
 		
 		$this->addClass('InputfieldNoFocus', 'wrapClass');
 		if(!$renderValueMode) $this->addClass('InputfieldHasUpload', 'wrapClass');
+		// When noAjax is set, advertise it on the wrapper so the client-side
+		// InputfieldFile.js skips HTML5 ajax initialization for this inputfield
+		// even when running in Page Edit (where #PageIDIndicator would
+		// otherwise force ajax for every InputfieldFile on the form).
+		// See: processwire/processwire-issues#1836
+		if($this->noAjax) {
+			$this->addClass('InputfieldNoAjaxUpload', 'wrapClass');
+		}
 		
 		if($this->useTags) {
 			$jQueryUI = $this->wire()->modules->get('JqueryUI'); /** @var JqueryUI $jQueryUI */
--- a/wire/modules/Inputfield/InputfieldFile/InputfieldFile.js
+++ b/wire/modules/Inputfield/InputfieldFile/InputfieldFile.js
@@ -203,6 +203,11 @@ function InputfieldFile($) {
 				$target.closest('.InputfieldContent').each(function (i) {
 					if($(this).hasClass('InputfieldFileInit')) return;
+					// Skip inputfields that have noAjax enabled — they fall back to
+					// InitOldSchool's delegated change handler and native form submit.
+					// See: processwire/processwire-issues#1836
+					if($(this).closest('.Inputfield').hasClass('InputfieldNoAjaxUpload')) return;
 					initHTML5Item($(this), i);
 					$(this).addClass('InputfieldFileInit');
 				});
@@ -603,11 +608,18 @@ function InputfieldFile($) {
 	 *
 	 * #PageIDIndictator.length indicates PageEdit, which we're limiting AjaxUpload to since only ProcessPageEdit has the ajax handler
 	 * 
 	 */
 	var allowAjax = false;
+	// Always bind InitOldSchool's delegated change handler. For HTML5-enabled
+	// inputfields the HTML5 change handler calls stopPropagation(), preventing
+	// the delegated handler from firing, so this is a no-op for them. For
+	// noAjax inputfields (marked with .InputfieldNoAjaxUpload) this is the
+	// only change handler that fires — providing UI label updates while
+	// leaving the actual file upload to the native form submit.
+	// See: processwire/processwire-issues#1836
+	InitOldSchool();
 	if (window.File && window.FileList && window.FileReader 
 			&& ($("#PageIDIndicator").length > 0 || $('.InputfieldAllowAjaxUpload').length > 0)) {  
 		InitHTML5('');  
 		allowAjax = true;
 	} else {
-		InitOldSchool();
+		// InitOldSchool() already called above
 	}
```

### What the fix does

**PHP side — `InputfieldFile::renderReady()`**
- When `$this->noAjax` is truthy, add the class `InputfieldNoAjaxUpload` to the inputfield's `wrapClass`. This is the natural inverse of the existing `InputfieldAllowAjaxUpload` marker class that `ProcessProfile` already uses to opt a non-Page-Edit form INTO ajax. `InputfieldNoAjaxUpload` opts a Page-Edit form OUT of ajax.

**JS side — `InitHTML5()`**
- Inside the per-`InputfieldContent` loop, skip any inputfield whose nearest `.Inputfield` ancestor has the `InputfieldNoAjaxUpload` class. Skipped fields don't get the HTML5 change listener (with its `preventDefault` + `this.value = ''`), so the native file input behaves normally and the form POST carries the file when the user saves.

**JS side — main init**
- Always call `InitOldSchool()` upfront, before deciding whether to also call `InitHTML5('')`. This binds the delegated `change` handler on `document` for `.InputfieldFileUpload input[type=file]`. For HTML5-initialized fields the HTML5 element-level change handler calls `evt.stopPropagation()`, so the delegated handler doesn't fire on them — making this a safe no-op for ajax-enabled fields. For `noAjax` fields it's now the only change handler that fires, and it just updates the file label and validates size — the actual upload still happens via the native form submit.

### Behavior after the fix

- `InputfieldFile` rendered in **Page Edit** with `noAjax=true` → the HTML5 change handler is NOT bound, the file input behaves like a normal HTML file input, the file is POSTed with the form on Save, and the user's hook can intercept it.
- `InputfieldFile` rendered in **Page Edit** with `noAjax=false` (default) → unchanged: HTML5 ajax upload still works.
- `InputfieldFile` rendered in a **non-Page-Edit form** with `noAjax=true` → unchanged: the `else` branch was already calling `InitOldSchool()`.
- `InputfieldFile` rendered in a **non-Page-Edit form** without any `InputfieldAllowAjaxUpload` field → unchanged: same `else`-branch behavior.

## Workaround without a core patch
```php
<?php namespace ProcessWire;

// Force a hook-added InputfieldFile in Page Edit to skip ajax upload init
$wire->addHookAfter('InputfieldFile::renderReady', function(HookEvent $event) {
    $inputfield = $event->object;
    if($inputfield->noAjax) {
        $inputfield->addClass('InputfieldNoAjaxUpload', 'wrapClass');
    }
});
```

That alone is enough for the **PHP-side half** of the fix — but the JS side also needs to know about the new class. If you also need the JS behavior immediately, you can ship a small site-side JS override in your admin theme's `admin.js` or via `$config->js('InputfieldFile', [...])`:

```js
// site/templates/admin.js  (or any JS file loaded in the admin)
$(document).on('ready reloaded', function() {
    // Mark noAjax inputfields as already-initialized so InitHTML5 skips them
    $('.InputfieldNoAjaxUpload .InputfieldContent').addClass('InputfieldFileInit');
});
```

That blocks InitHTML5 from binding its `change` handler on the noAjax inputfields. The native file input then works as expected (form submit carries the file). The label-update behavior of `InitOldSchool`'s delegated handler may or may not be bound depending on whether `allowAjax` was true at page load — if not, just call `InitOldSchool()` yourself in your site JS.
